cpp/src/io/orc/reader_impl.cu

```diff
@@ -683,6 +671,88 @@ rmm::device_buffer reader::impl::decompress_stripe_data(
+                      rmm::cuda_stream_view stream,
+                      rmm::mr::device_memory_resource* mr)
+{
+  const auto num_stripes = chunks.size().first;
+  const auto num_columns = chunks.size().second;
+  bool is_mask_updated   = false;
+
+  for (size_t j = 0; j < num_columns; ++j) {
+    if (chunks[0][j].parent_data.valid_map_base) {
+      if (not is_mask_updated) chunks.device_to_host(stream, true);
+
+      auto parent_valid_map_base = chunks[0][j].parent_data.valid_map_base;
+      auto child_valid_map_base  = out_buffers[j].null_mask();
+      auto child_mask_len  = chunks[0][j].column_num_rows - chunks[0][j].parent_data.null_count;
+      auto parent_mask_len = chunks[0][j].column_num_rows;
+
+      if (child_valid_map_base) {
+        rmm::device_uvector<uint32_t> dst_idx(child_mask_len, stream);
+        // Copy indexes at which the parent has valid value.
+        thrust::copy_if(rmm::exec_policy(stream),
+                        thrust::make_counting_iterator(0),
+                        thrust::make_counting_iterator(0) + parent_mask_len,
+                        dst_idx.begin(),
+                        [parent_valid_map_base] __device__(auto idx) {
+                          auto val = bit_is_set(parent_valid_map_base, idx);
+                          return val;
+                        });
+
+        auto merged_null_mask = cudf::detail::create_null_mask(
+          parent_mask_len, mask_state::ALL_NULL, rmm::cuda_stream_view(stream), mr);
+        bitmask_type* merged_mask = static_cast<bitmask_type*>(merged_null_mask.data());
+        uint32_t* dst_idx_ptr     = dst_idx.data();
+        // Copy child valid bits from child column to valid indexes, this will merge both child and
+        // parent null masks
+        thrust::for_each(rmm::exec_policy(stream),
+                         thrust::make_counting_iterator(0),
+                         thrust::make_counting_iterator(0) + dst_idx.size(),
+                         [child_valid_map_base, dst_idx_ptr, merged_mask] __device__(auto idx) {
+                           if (bit_is_set(child_valid_map_base, idx)) {
+                             cudf::set_bit(merged_mask, dst_idx_ptr[idx]);
+                           };
+                         });
+
+        out_buffers[j]._null_mask = std::move(merged_null_mask);
+
+      } else {
+        // Since child column doesn't have a mask, copy parent null mask
+        auto mask_size            = bitmask_allocation_size_bytes(parent_mask_len);
+        out_buffers[j]._null_mask = std::move(
+          rmm::device_buffer(static_cast<void*>(parent_valid_map_base), mask_size, stream, mr));
+      }
+      if (not is_mask_updated) is_mask_updated = true;
+    }
+  }
+  if (is_mask_updated) {
+    // Update chunks with pointers to column data which might have been changed.
+    for (size_t i = 0; i < num_stripes; ++i) {
+      for (size_t j = 0; j < num_columns; ++j) {
```

Comment:
here again I would prefer stripe_idx and col_idx, but in this case it's much less relevant since the loop is very simple.